<br>

<div style="text-align: center">Confidential – Qualcomm Technologies, Inc. and/or its affiliated companies – May Contain Trade Secrets</div>

<div style="text-align: center; font-weight: bold">MAY CONTAIN U.S. AND INTERNATIONAL EXPORT CONTROLLED INFORMATION</div> <br>

**Takeaways:** Users will learn how to tune the Resnet50 model for best throughput and latency

**Before you start:** 
- There are some commands (folder locations etc) that will need to be updated in this notebook based on the platform and installation location. Some commands might need sudo prefix to run properly.
- The terms 'model' and 'network' are used interchangeably in this notebook. 
- The terms 'NSP' and 'AI compute core" are used interchangeably in this notebook.

**Last Verified Qualcomm Cloud AI Platform SDK and Apps SDK Version:** Platform SDK 1.9.2.14 and Apps SDK 1.9.1.25 

# <span style='color:Blue'> Performance Tuning on Cloud AI </span>

##  Pre-requisite reading 
New users on Cloud AI platforms are expected to go over the Cloud AI SoC architecture and the key compile/runtime parameters that determine performance. 


## Introduction 
This notebook is for beginners and will take the user through the workflow to achieve best throughput and latency on Cloud AI platforms for the Resnet50 model from HuggingFace. 

Here is the workflow that will be demonstated in this notebook. 

1. **Install required packages**: Begin by installing all the required packages
2. **Import the model**: Download the Resnet50 model in ONNX. 
2. **Device Health Check**: Query the device health using qaic-util tool. 
2. **Identify best throughput configuration**: Go over the Model Configurator tool to identify best throughput 
3. **Identity least latency configuration**: Go over the key parameters to tweak for least latency


# <span style='color:Blue'> 1. Install required packages </span>

We will install the required Python packages 

In [2]:
!pip install -r requirements.txt

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.1/10.1 MB 114.5 MB/s eta 0:00:00a 0:00:01
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.1/17.1 MB 115.2 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 890.2/890.2 MB 1.4 MB/s eta 0:00:0000:0100:18
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.8/48.8 MB 110.8 MB/s eta 0:00:00a 0:00:01
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.7/61.7 MB 115.2 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 224.8/224.8 kB 281.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 162.1/162.1 kB 161.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 849.3/849.3 kB 146.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 557.1/557.1 MB 115.4 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.1/317.1 MB 114.9 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 115.2 MB/s eta 0:00:00a 0:00:01
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtracking for guidance. If you want to abort this run, press Ctrl + C.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 243.9/243.9 kB 455.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 318.1/318.1 kB 115.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 318.1/318.1 kB 116.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 243.7/243.7 kB 136.8 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 243.7/243.7 kB 113.0 MB/s eta 0:

  Preparing metadata (pyproject.toml) ... done
ERROR: Cannot install -r requirements.txt (line 1), -r requirements.txt (line 4), numpy==1.23.4 and onnxruntime==1.15.1 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested numpy==1.23.4
    onnx 1.12.0 depends on numpy>=1.16.6
    onnxruntime 1.15.1 depends on numpy>=1.24.2
    The user requested numpy==1.23.4
    onnx 1.12.0 depends on numpy>=1.16.6
    onnxruntime 1.15.0 depends on numpy>=1.24.2

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts

[notice] A new release of pip is available: 23.1.2 -> 23.2.1
[notice] To update, run: pip install --upgrade pip


# <span style='color:Blue'>2. Download the model </span>

Download the pretrained Resnet50 model. 

In [23]:
!wget https://zenodo.org/record/2592612/files/resnet50_v1.onnx

--2023-08-03 16:42:53--  https://zenodo.org/record/2592612/files/resnet50_v1.onnx
Resolving zenodo.org (zenodo.org)... 188.185.124.72
Connecting to zenodo.org (zenodo.org)|188.185.124.72|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 102164585 (97M) [application/octet-stream]
Saving to: ‘resnet50_v1.onnx.1’

resnet50_v1.onnx.1  100%[===================>]  97.43M  5.48MB/s    in 52s     

2023-08-03 16:43:46 (1.88 MB/s) - ‘resnet50_v1.onnx.1’ saved [102164585/102164585]



In [9]:
!source /opt/qti-aic/dev/python/qaic-env/bin/activate
import sys
assert sys.executable  == '/opt/qti-aic/dev/python/qaic-env/bin/python'

!sudo {sys.executable} -W ignore /opt/qti-aic/tools/qaic-pytools/qaic-model-preparator.py --config resnet50_model_info.yaml

Badly placed ()'s.
2023-08-03 22:04:27.945165: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2023-08-03 22:04:27.945186: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2023-08-03 22:04:28.951 | WARNING  | __main__:prepare_work_dir:39 - Directory WORKSPACE_RESNET50 will be deleted if already exists. Take backup before execution.
2023-08-03 22:04:28.954 | INFO     | __main__:main:72 - QAic Model Preparator Tool
2023-08-03 22:04:29.016 | INFO     | qaic_pytools.qmodel.backend.onnx.model:native_checker:753 - Graph checker passed successfully.
2023-08-03 22:04:29.080 | INFO     | qaic_pytools.qmodel.preparator.preparator:__prepare_onnx:283 - Native Checker on the input model is successful
2023-08-03 22:04:29.080 | INFO     | qaic_pytools.qmodel.preparator.

# <span style='color:Blue'> 3. Check device health </span>

'qaic-util' tool can be used to query the health of all the Cloud AI cards in the server. 

"Status: Ready" indicates that the health of the cards is good. 
"Status: Error" indicates that the cards are not in good health and a system administrator needs to be contacted to rectify the issue. 

In [1]:
!/opt/qti-aic/tools/qaic-util -q | grep "Status"

	Status:Ready


# <span style='color:Blue'> 4. Identify the best throughput configuration </span>

Model Configurator is a python script that is used to find the optimal configuration of batch size, cores etc for a given model for the throughput (Inf/s). The input to model configurator is the model and a search space that the tool iterates over.   

1. Use the optimal config (model configurator output) as an indicator of the best compile flags.
2. Compile with qaic-exec and run with qaic-runner using optimal config from step 1 to validate the performance. 

Key compile optimization flags are 
- **aic-num-cores** : # of AI compute cores (aka NSP) used to compile the model
- **bs** : batch size
- **mos** : Maximum output splitting, denotes the no of AI compute cores across which an output channel (and associated weights) is split. 
- **ols** : Overlap splitting, enabled output splitting to improve core level parallelism, eg tensor and vector unit

Key runtime optimization flags are 
- **instance/activations** : No of instances of the compiled binary that can be run based on the # of AI compute cores on the card 
- **set-size** : denotes the number of inferences per instance that can be queued up on the host. Hides host side overhead by pipelining inferences 

## Config Optimizer 
The optimized search is run on one or more searchable parameters (aic-num-cores, instances, batch-size etc). The search space is provided through a json configuration file. 
This table captures the key elements of the configuration file. 

| Key | Type | Description | Recommended value|
| :- |:- | :- | :-|
|"max_func_eval"|Integer|Maximun number of evaluations to do for each initial point. This number can be increased if successful convergence is not achieved|200|
|"Objective"|String| Search objective. Options are "maximize_inf_rate" or "minimize_latency" | Choose "maximize_inf_rate" for maximum throughput or "minimize_latency" for minimum latency |
|“params”|Json Object|Provide the search range for each of the parameters - cores, mos, ols, etc. through min, max values| See table below |
“initial_values”|List of Json Objects|List of initial values for the search parameters. A fresh search is initiated from each of these points and the results returned. Initial Values must be picked from within the search range defined in “params”|	Provide multiple initial values as shown in the example json|
|“static_params”|Json Object|Optional static values to be used for searchable parameters which have been excluded from the search space| |


### Parameter Range


| Parameter|	Recommended Search Range|	Valid Range |	Comments|
| :- | :- | :- | :-|
|cores|	1-Number of NSP on device|	1-Number of NSP on device||	
|mos	|1-Number of NSP on device|	1-Number of NSP on device||	
|ols	|1-8|	Integers>0|	
|batch-size (bs)|	1-16|	Integers>0|	Min, max values must be power of 2. The max value would depend on the model|
|instances|	1-Number of NSP on device|	1-Number of NSP on device|	

In [17]:
#Lets assume the max Number of NSP on the device is 14. 
!cat resnet_base_dopt_throughput.json































































}

In [5]:
import sys
assert sys.executable  == '/opt/qti-aic/dev/python/qaic-env/bin/python'

!{sys.executable} /opt/qti-aic/scripts/qaic-model-configurator/model_configurator.py WORKSPACE_RESNET50/resnet50_v1_preparator_aic100.onnx onnx \
-quantization-schema-activations=symmetric_with_uint8 -quantization-schema-constants=symmetric_with_uint8 \
-quantization-precision=Int8 -multicast-weights \
-optimized-config-search resnet_base_dopt.json -max-compilation-threads 8 -time 5 \
-device-id 0 

/opt/qti-aic/dev/python/qaic-env/lib/python3.8/site-packages/paramiko/transport.py:236: CryptographyDeprecationWarning: Blowfish has been deprecated
  "class": algorithms.Blowfish,
2023-08-03 11:11:20.530 - [INFO]: Starting /opt/qti-aic/scripts/qaic-model-configurator/model_configurator.py resnet50_v1_preparator_aic100.onnx onnx -quantization-schema-activations=symmetric_with_uint8 -quantization-schema-constants=symmetric_with_uint8 -quantization-precision=Int8 -multicast-weights -optimized-config-search resnet_base_dopt.json -max-compilation-threads 8 -time 5 -device-id 0
2023-08-03 11:11:20.530 - [INFO]: Model Name: resnet50_v1_preparator_aic100.onnx
2023-08-03 11:11:20.531 - [INFO]: Hostname: rpuvvada-linux, Physical Cores: 8, Logical Cores: 16, Memory: 31.1 GB
2023-08-03 11:11:20.616 - [INFO]: Running optimized search
[2023-08-03 11:11:20.644] [info] Compiling model with compiler parameters: [(cores=1, mos=[1], ols=1, batchSize=1)]
W0803 11:11:24.548735 2862225 Quantization.cpp:422

[2023-08-03 11:16:44.454] [info] Compiling model with compiler parameters: [(cores=1, mos=[1], ols=7, batchSize=16)]
W0803 11:16:44.540076 2874081 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
[2023-08-03 11:17:20.105] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-7-bs-16-output with runnerParams (instances=14)
[2023-08-03 11:17:23.150] [info] Model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-7-bs-16-output with runnerParams (instances=14) running on device ID 0
[2023-08-03 11:17:29.395] [info] Compiling model with compiler parameters: [(cores=1, mos=[1], ols=4, batchSize=8)]
W0803 11:17:29.487104 2874720 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
[2023-08-03 11:18:04.732] [info] Running model at path model_configurator_output/compiled_mo

[2023-08-03 11:21:43.387] [info] Compiling model with compiler parameters: [(cores=2, mos=[1], ols=1, batchSize=8), (cores=2, mos=[1], ols=2, batchSize=4), (cores=2, mos=[2], ols=1, batchSize=4)]
W0803 11:21:43.494024 2884397 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
W0803 11:21:43.547706 2884399 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
W0803 11:21:43.549463 2884398 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
[2023-08-03 11:22:21.853] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-1-bs-4-output with runnerParams (instances=6)
[2023-08-03 11:22:22.868] [info] Model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-1-bs-4-output with runnerParams (instan

[2023-08-03 11:25:47.037] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-4-mos-3-ols-1-bs-1-output with runnerParams (instances=3)
[2023-08-03 11:25:47.547] [info] Model at path model_configurator_output/compiled_models/qpc-cores-4-mos-3-ols-1-bs-1-output with runnerParams (instances=3) running on device ID 0
[2023-08-03 11:25:52.601] [info] Compiling model with compiler parameters: [(cores=3, mos=[2], ols=1, batchSize=1), (cores=4, mos=[2], ols=1, batchSize=2), (cores=4, mos=[2], ols=2, batchSize=1)]
W0803 11:25:52.731000 2887422 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
W0803 11:25:52.761622 2887421 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
W0803 11:25:52.762609 2887423 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend

[2023-08-03 11:31:12.414] [info] Model at path model_configurator_output/compiled_models/qpc-cores-4-mos-4-ols-1-bs-4-output with runnerParams (instances=2) running on device ID 0
[2023-08-03 11:31:17.457] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-4-mos-4-ols-2-bs-4-output with runnerParams (instances=3)
[2023-08-03 11:31:18.227] [info] Model at path model_configurator_output/compiled_models/qpc-cores-4-mos-4-ols-2-bs-4-output with runnerParams (instances=3) running on device ID 0
[2023-08-03 11:31:23.281] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-4-mos-4-ols-1-bs-8-output with runnerParams (instances=3)
[2023-08-03 11:31:24.592] [info] Model at path model_configurator_output/compiled_models/qpc-cores-4-mos-4-ols-1-bs-8-output with runnerParams (instances=3) running on device ID 0
[2023-08-03 11:31:30.159] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-2-bs-4-outp

[2023-08-03 11:37:38.328] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-3-mos-3-ols-2-bs-16-output with runnerParams (instances=2)
[2023-08-03 11:37:40.701] [info] Model at path model_configurator_output/compiled_models/qpc-cores-3-mos-3-ols-2-bs-16-output with runnerParams (instances=2) running on device ID 0
[2023-08-03 11:37:46.262] [info] Compiling model with compiler parameters: [(cores=4, mos=[2], ols=2, batchSize=8)]
W0803 11:37:46.360864 2899005 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
[2023-08-03 11:38:26.521] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-4-mos-2-ols-2-bs-8-output with runnerParams (instances=3)
[2023-08-03 11:38:27.884] [info] Model at path model_configurator_output/compiled_models/qpc-cores-4-mos-2-ols-2-bs-8-output with runnerParams (instances=3) running on device ID 0
[2023-08-03 11:38:33.453] [in

[2023-08-03 11:43:34.466] [info] Compiling model with compiler parameters: [(cores=5, mos=[1], ols=1, batchSize=1)]
W0803 11:43:34.572508 2901437 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
[2023-08-03 11:44:04.237] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-5-mos-1-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 11:44:04.729] [info] Model at path model_configurator_output/compiled_models/qpc-cores-5-mos-1-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 11:44:09.774] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-6-mos-1-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 11:44:10.124] [info] Model at path model_configurator_output/compiled_models/qpc-cores-6-mos-1-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 11:44:15.167] [info

[2023-08-03 11:51:05.233] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-13-mos-2-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 11:51:05.748] [info] Model at path model_configurator_output/compiled_models/qpc-cores-13-mos-2-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 11:51:10.789] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-13-mos-1-ols-2-bs-1-output with runnerParams (instances=1)
[2023-08-03 11:51:11.173] [info] Model at path model_configurator_output/compiled_models/qpc-cores-13-mos-1-ols-2-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 11:51:16.209] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-13-mos-1-ols-1-bs-2-output with runnerParams (instances=1)
[2023-08-03 11:51:16.960] [info] Model at path model_configurator_output/compiled_models/qpc-cores-13-mos-1-ols-1-bs-2-output with runnerPar

[2023-08-03 11:58:05.987] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-9-mos-1-ols-1-bs-16-output with runnerParams (instances=1)
[2023-08-03 11:58:08.475] [info] Model at path model_configurator_output/compiled_models/qpc-cores-9-mos-1-ols-1-bs-16-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 11:58:13.507] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-10-mos-1-ols-1-bs-8-output with runnerParams (instances=1)
[2023-08-03 11:58:14.869] [info] Model at path model_configurator_output/compiled_models/qpc-cores-10-mos-1-ols-1-bs-8-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 11:58:19.904] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-11-mos-1-ols-1-bs-16-output with runnerParams (instances=1)
[2023-08-03 11:58:22.184] [info] Model at path model_configurator_output/compiled_models/qpc-cores-11-mos-1-ols-1-bs-16-output with runnerP

## Measure end-to-end latency 

Here are the steps to measure the end-to-end latency as well as the breakdown of latency for the best configuration identified in the previous step. 

1. Compile the model using 'qaic-exec'with the configuration parameters identified in previous step using model configurator. 
2. Execute the compiled model using 'qaic-runner' with the run time parameters identified in previous step using model configurator. Run 'qaic-runner' with flags that dump latency information.
3. Post process the latency information to identify percentile distribution (mean, median, 95 and 99) in latency across inferences.

In [6]:
## Compile the model 

!rm -rf compiled_fp16
!/opt/qti-aic/exec/qaic-exec -v -aic-hw  \
-quantization-schema-activations=symmetric_with_uint8 -quantization-schema-constants=symmetric_with_uint8 \
-mos=1 -ols=1 -aic-num-cores=2 -batchsize=4 \
-m=WORKSPACE_RESNET50/resnet50_v1_preparator_aic100.onnx \
-stats-batchsize=4 -aic-binary-dir=./compiled_int8 \
-aic-hw-version=2.0 -compile-only

Reading ONNX Model from resnet50_v1_preparator_aic100.onnx
Compile started ............... 
Compiling model with Int8 precision using PGQ.
W0803 12:13:53.626268 2932560 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
Generated binary is present at ./compiled_int8


In [15]:
## Execute the compiled model with the latency flags 
!mkdir resnet50_cased_stats
!/opt/qti-aic/exec/qaic-runner --test-data ./compiled_int8  -d 0 -a 7 \
--aic-profiling-type latency --aic-profiling-out-dir ./resnet50_cased_stats \
--aic-profiling-start-iter 100 --aic-profiling-num-samples 99999 --time 20 

mkdir: cannot create directory ‘resnet50_cased_stats’: File exists
 ---- Stats ----
InferenceCnt 89241 TotalDuration 20015187us BatchSize 1 Inf/Sec 4458.664
Deleting previous file: ./resnet50_cased_stats/aic-profiling-program-0-latency.txt
Writing file:./resnet50_cased_stats/aic-profiling-program-0-latency.txt
Deleting previous file: ./resnet50_cased_stats/aic-profiling-program-0-latency.txt
Writing file:./resnet50_cased_stats/aic-profiling-program-0-latency.txt
Deleting previous file: ./resnet50_cased_stats/aic-profiling-program-0-latency.txt
Writing file:./resnet50_cased_stats/aic-profiling-program-0-latency.txt
Deleting previous file: ./resnet50_cased_stats/aic-profiling-program-0-latency.txt
Writing file:./resnet50_cased_stats/aic-profiling-program-0-latency.txt
Deleting previous file: ./resnet50_cased_stats/aic-profiling-program-0-latency.txt
Writing file:./resnet50_cased_stats/aic-profiling-program-0-latency.txt
Deleting previous file: ./resnet50_cased_stats/aic-profiling-program

### Latency breakdown 

The end-to-end inference latency can be broken down into 4 major categories - Application, Linux Runtime (LRT) processing, Kernel mode driver (KMD) processing and Cloud AI device processing. 


![Latency Breakdown](Images/Latency.jpg)

**Key Latency Stats** (units in us)
- *totalRoundtripTime* :Time from point where application (qaic-runner in this case) calls Runtime API and ends where post-processing is complete and control is returned to the application indicating the inference is complete
- *preProcTime* : Time taken to pre-process the data (model input) on the host
- *postProc* : Time taken to post-process the data (model output) on the host
- *execTotal* : Time from inference object being submitted to kernel, completion on hardware and processing is returned to user-space


In [16]:
## Post process the latency information to identify percentile distribution

!{sys.executable} latency_stats_python3.py ./resnet50_cased_stats/aic-profiling-program-0-latency.txt 

All activations combined:
                         mean        min        50%        75%        90%        95%        99%        max
hostRoundTrip       15.587172  15.026000  15.590000  15.688000  15.726000  15.829000  15.915000  16.313000
enqTime              0.001774   0.000888   0.001482   0.002067   0.002667   0.002984   0.004933   0.011712
preProcTime          0.073094   0.046538   0.069450   0.079348   0.099532   0.113489   0.131796   0.174087
submitTime           0.000472   0.000286   0.000448   0.000507   0.000597   0.000645   0.000774   0.010455
execTotal           15.604910  15.031588  15.610173  15.709955  15.753528  15.846528  15.934004  16.342660
exectoVc             0.001849   0.001000   0.002000   0.002000   0.002000   0.002000   0.003000   0.009000
execToComplete      15.585323  15.025000  15.588000  15.686000  15.724000  15.827000  15.913000  16.311000
postProc             0.001046   0.000410   0.000980   0.001281   0.001516   0.001666   0.002075   0.009754
totalRoundt

As observed, for the configuration choosen, throughput of 15940 inf/sec is observed with a latency of 16.023 ms

# <span style='color:Blue'> 5. Identify the least latency configuration </span>

Identifying the least latency configurations requires the users to run the model_configurator tool with the "objective" parameter set to "minimize_latency". 


Lets go through these steps for the resnet50 model.

In [22]:
!{sys.executable} /opt/qti-aic/scripts/qaic-model-configurator/model_configurator.py WORKSPACE_RESNET50/resnet50_v1_preparator_aic100.onnx onnx \
-quantization-schema-activations=symmetric_with_uint8 -quantization-schema-constants=symmetric_with_uint8 \
-quantization-precision=Int8 -multicast-weights \
-optimized-config-search resnet_base_dopt_min_latency.json -max-compilation-threads 8 -time 5 \
-device-id 0 -set-size 1

/opt/qti-aic/dev/python/qaic-env/lib/python3.8/site-packages/paramiko/transport.py:236: CryptographyDeprecationWarning: Blowfish has been deprecated
  "class": algorithms.Blowfish,
2023-08-03 16:08:58.535 - [INFO]: Starting /opt/qti-aic/scripts/qaic-model-configurator/model_configurator.py resnet50_v1_preparator_aic100.onnx onnx -quantization-schema-activations=symmetric_with_uint8 -quantization-schema-constants=symmetric_with_uint8 -quantization-precision=Int8 -multicast-weights -optimized-config-search resnet_base_dopt_min_latency.json -max-compilation-threads 8 -time 5 -device-id 0 -set-size 1
2023-08-03 16:08:58.535 - [INFO]: Model Name: resnet50_v1_preparator_aic100.onnx
2023-08-03 16:08:58.536 - [INFO]: Hostname: rpuvvada-linux, Physical Cores: 8, Logical Cores: 16, Memory: 31.1 GB
2023-08-03 16:08:58.622 - [INFO]: Running optimized search
[2023-08-03 16:08:58.657] [info] Compiling model with compiler parameters: [(cores=1, mos=[1], ols=1, batchSize=1)]
W0803 16:09:02.557143 3117

[2023-08-03 16:11:58.328] [info] Model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-3-bs-1-output with runnerParams (instances=14) running on device ID 0
[2023-08-03 16:12:03.474] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-3-bs-1-output with runnerParams (instances=14)
[2023-08-03 16:12:04.179] [info] Model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-3-bs-1-output with runnerParams (instances=14) running on device ID 0
[2023-08-03 16:12:09.329] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-3-bs-1-output with runnerParams (instances=14)
[2023-08-03 16:12:10.004] [info] Model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-3-bs-1-output with runnerParams (instances=14) running on device ID 0
[2023-08-03 16:12:15.155] [info] Compiling model with compiler parameters: [(cores=1, mos=[1], ols=2, batchSize=2)]
W0803 16:

[2023-08-03 16:15:13.641] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-1-bs-1-output with runnerParams (instances=7)
[2023-08-03 16:15:14.182] [info] Model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-1-bs-1-output with runnerParams (instances=7) running on device ID 0
[2023-08-03 16:15:19.277] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-2-mos-1-ols-1-bs-1-output with runnerParams (instances=6)
[2023-08-03 16:15:19.814] [info] Model at path model_configurator_output/compiled_models/qpc-cores-2-mos-1-ols-1-bs-1-output with runnerParams (instances=6) running on device ID 0
[2023-08-03 16:15:24.897] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-2-mos-2-ols-1-bs-1-output with runnerParams (instances=7)
[2023-08-03 16:15:25.423] [info] Model at path model_configurator_output/compiled_models/qpc-cores-2-mos-2-ols-1-bs-1-output with runnerParams (i

[2023-08-03 16:17:24.148] [info] Compiling model with compiler parameters: [(cores=4, mos=[1], ols=1, batchSize=1)]
W0803 16:17:24.264390 3133116 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
[2023-08-03 16:17:53.672] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-4-mos-1-ols-1-bs-1-output with runnerParams (instances=3)
[2023-08-03 16:17:54.119] [info] Model at path model_configurator_output/compiled_models/qpc-cores-4-mos-1-ols-1-bs-1-output with runnerParams (instances=3) running on device ID 0
[2023-08-03 16:17:59.173] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-4-mos-1-ols-1-bs-1-output with runnerParams (instances=3)
[2023-08-03 16:17:59.584] [info] Model at path model_configurator_output/compiled_models/qpc-cores-4-mos-1-ols-1-bs-1-output with runnerParams (instances=3) running on device ID 0
[2023-08-03 16:18:04.639] [info

[2023-08-03 16:20:21.431] [info] Model at path model_configurator_output/compiled_models/qpc-cores-3-mos-1-ols-1-bs-1-output with runnerParams (instances=3) running on device ID 0
[2023-08-03 16:20:26.731] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-4-mos-1-ols-1-bs-1-output with runnerParams (instances=2)
[2023-08-03 16:20:27.193] [info] Model at path model_configurator_output/compiled_models/qpc-cores-4-mos-1-ols-1-bs-1-output with runnerParams (instances=2) running on device ID 0
[2023-08-03 16:20:32.235] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-4-mos-2-ols-1-bs-1-output with runnerParams (instances=3)
[2023-08-03 16:20:32.736] [info] Model at path model_configurator_output/compiled_models/qpc-cores-4-mos-2-ols-1-bs-1-output with runnerParams (instances=3) running on device ID 0
[2023-08-03 16:20:37.788] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-4-mos-1-ols-2-bs-1-outp

[2023-08-03 16:22:53.791] [info] Model at path model_configurator_output/compiled_models/qpc-cores-3-mos-1-ols-2-bs-1-output with runnerParams (instances=2) running on device ID 0
[2023-08-03 16:22:58.833] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-3-mos-1-ols-1-bs-2-output with runnerParams (instances=2)
[2023-08-03 16:22:59.577] [info] Model at path model_configurator_output/compiled_models/qpc-cores-3-mos-1-ols-1-bs-2-output with runnerParams (instances=2) running on device ID 0
[2023-08-03 16:23:04.619] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-2-mos-1-ols-1-bs-1-output with runnerParams (instances=2)
[2023-08-03 16:23:05.093] [info] Model at path model_configurator_output/compiled_models/qpc-cores-2-mos-1-ols-1-bs-1-output with runnerParams (instances=2) running on device ID 0
[2023-08-03 16:23:10.128] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-3-mos-1-ols-1-bs-1-outp

[2023-08-03 16:25:01.305] [info] Model at path model_configurator_output/compiled_models/qpc-cores-3-mos-1-ols-1-bs-2-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 16:25:06.336] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-4-mos-1-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 16:25:06.773] [info] Model at path model_configurator_output/compiled_models/qpc-cores-4-mos-1-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 16:25:11.801] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-3-mos-2-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 16:25:12.292] [info] Model at path model_configurator_output/compiled_models/qpc-cores-3-mos-2-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 16:25:17.325] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-3-mos-1-ols-2-bs-1-outp

[2023-08-03 16:29:37.259] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-8-mos-1-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 16:29:37.546] [info] Model at path model_configurator_output/compiled_models/qpc-cores-8-mos-1-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 16:29:42.573] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-7-mos-2-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 16:29:43.066] [info] Model at path model_configurator_output/compiled_models/qpc-cores-7-mos-2-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 16:29:48.100] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-7-mos-1-ols-2-bs-1-output with runnerParams (instances=1)
[2023-08-03 16:29:48.598] [info] Model at path model_configurator_output/compiled_models/qpc-cores-7-mos-1-ols-2-bs-1-output with runnerParams (i

[2023-08-03 16:31:43.738] [info] Compiling model with compiler parameters: [(cores=14, mos=[1], ols=1, batchSize=1)]
W0803 16:31:43.856567 3144723 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
[2023-08-03 16:32:42.043] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-14-mos-1-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 16:32:42.358] [info] Model at path model_configurator_output/compiled_models/qpc-cores-14-mos-1-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 16:32:47.394] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-14-mos-1-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 16:32:47.755] [info] Model at path model_configurator_output/compiled_models/qpc-cores-14-mos-1-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 16:32:52.789] 

[2023-08-03 16:35:23.573] [info] Model at path model_configurator_output/compiled_models/qpc-cores-7-mos-2-ols-1-bs-1-output with runnerParams (instances=2) running on device ID 0
[2023-08-03 16:35:28.615] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-7-mos-1-ols-2-bs-1-output with runnerParams (instances=2)
[2023-08-03 16:35:29.048] [info] Model at path model_configurator_output/compiled_models/qpc-cores-7-mos-1-ols-2-bs-1-output with runnerParams (instances=2) running on device ID 0
[2023-08-03 16:35:34.088] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-7-mos-1-ols-1-bs-2-output with runnerParams (instances=2)
[2023-08-03 16:35:34.623] [info] Model at path model_configurator_output/compiled_models/qpc-cores-7-mos-1-ols-1-bs-2-output with runnerParams (instances=2) running on device ID 0
[2023-08-03 16:35:39.672] [info] Compiling model with compiler parameters: [(cores=13, mos=[1], ols=1, batchSize=1), (cores=14, mo

[2023-08-03 16:39:26.352] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-12-mos-1-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 16:39:26.815] [info] Model at path model_configurator_output/compiled_models/qpc-cores-12-mos-1-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 16:39:31.848] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-12-mos-1-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 16:39:32.304] [info] Model at path model_configurator_output/compiled_models/qpc-cores-12-mos-1-ols-1-bs-1-output with runnerParams (instances=1) running on device ID 0
[2023-08-03 16:39:37.335] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-12-mos-1-ols-1-bs-1-output with runnerParams (instances=1)
[2023-08-03 16:39:37.788] [info] Model at path model_configurator_output/compiled_models/qpc-cores-12-mos-1-ols-1-bs-1-output with runnerPar

[2023-08-03 16:42:40.773] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-1-bs-1-output with runnerParams (instances=8)
[2023-08-03 16:42:41.410] [info] Model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-1-bs-1-output with runnerParams (instances=8) running on device ID 0
[2023-08-03 16:42:46.509] [info] Running model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-1-bs-1-output with runnerParams (instances=8)
[2023-08-03 16:42:47.143] [info] Model at path model_configurator_output/compiled_models/qpc-cores-1-mos-1-ols-1-bs-1-output with runnerParams (instances=8) running on device ID 0
2023-08-03 16:42:52.314 - [INFO]: Optimized search results:
   OptimizerStatus                        InitialValue                      cores mos   ols  batchSize  instances  Objective (micro-sec)
0  NUMERICAL_FAILURE  (cores=1, mos=[1], ols=1, batchSize=1, instances=14)    1    [1]   2       1         13    

The Configuration with 13 instances and batch size 1 is returned as the best confing amongst the ranges selected in the config file.

In [13]:
## aic-num-cores =7 and instances = 2
!rm -rf compiled_int8
!rm -rf resnet50_cased_stats_latency

!/opt/qti-aic/exec/qaic-exec -v -aic-hw  \
-quantization-schema-activations=symmetric_with_uint8 -quantization-schema-constants=symmetric_with_uint8 \
-mos=1 -ols=2 -aic-num-cores=1 -batchsize=1 \
-m=WORKSPACE_RESNET50/resnet50_v1_preparator_aic100.onnx \
-stats-batchsize=1 -aic-binary-dir=./compiled_int8 \
-aic-hw-version=2.0 -compile-only  

!mkdir resnet50_cased_stats_latency
!/opt/qti-aic/exec/qaic-runner --test-data ./compiled_int8  -d 0 -S 1 -a 13 \
--aic-profiling-type latency --aic-profiling-out-dir ./resnet50_cased_stats_latency \
--aic-profiling-start-iter 100 --aic-profiling-num-samples 99999 --time 20 | grep "InferenceCnt"

!{sys.executable} latency_stats_python3.py ./resnet50_cased_stats_latency/aic-profiling-program-0-latency.txt | grep "totalRoundtripTime:"



Reading ONNX Model from resnet50_v1_preparator_aic100.onnx
Compile started ............... 
Compiling model with Int8 precision using PGQ.
W0803 22:15:47.770970 3398690 Quantization.cpp:422] WARNING: Found INT8 unsupported node softmax_tensor/ in QAicGraph for AIC backend. Running in higher precision
Generated binary is present at ./compiled_int8
InferenceCnt 116640 TotalDuration 20001893us BatchSize 1 Inf/Sec 5831.448
Deleting previous file: ./resnet50_cased_stats_latency/aic-profiling-program-0-latency.txt
Deleting previous file: ./resnet50_cased_stats_latency/aic-profiling-program-0-latency.txt
Deleting previous file: ./resnet50_cased_stats_latency/aic-profiling-program-0-latency.txt
Deleting previous file: ./resnet50_cased_stats_latency/aic-profiling-program-0-latency.txt
Deleting previous file: ./resnet50_cased_stats_latency/aic-profiling-program-0-latency.txt
Deleting previous file: ./resnet50_cased_stats_latency/aic-profiling-program-0-latency.txt
Deleting previous file: ./resne

Latency can be reduced to 2.302ms from 16.023ms by setting the flag "-S" to 1, Batchsize 1 and -aic-num-cores=1. However the throughput reduced from 15940 inf/sec to 5831 inf/sec.